File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-5.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        bool top_row = false;
        for (int c = 0; c < n; ++c) {
            if (matrix[0][c] == 0) {
                top_row = true;
                break;
            }
        }

        for (int r = 1; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (matrix[r][c] == 0) {
                    matrix[r][0] = 0;
                    matrix[0][c] = 0;
                }
            }
        }

        // cout << top_row << endl;
        // for (int r = 0; r < m; ++r) {
        //     for (int c = 0; c < n; ++c) {
        //         cout << matrix[r][c] << " ";
        //     }
        //     cout << endl;
        // }
        for (int r = 1; r < m; ++r) {
            if (matrix[r][0] == 0) {
                for (int c = 1; c < n; ++c) {
                    matrix[r][c] = 0;
                }
            }
        }
        
        for (int c = 0; c < n; ++c) {
            if (matrix[0][c] == 0) {
                for (int r = 1; r < m; ++r) {
                    matrix[r][c] = 0;
                }
            }
        }

        if (top_row) {
            for (int c = 0; c < n; ++c) {
                matrix[0][c] = 0;
            }
        }
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-5.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        res = []
        min_r = min_c = 0
        max_r = m - 1
        max_c = n - 1
        r = c = 0
        dr, dc = 0, 1
        while len(res) < m * n:
            res.append(matrix[r][c])
            if dc == 1 and c == max_c:
                min_r += 1
                dr, dc = 1, 0
            elif dc == -1 and c == min_c:
                max_r -= 1
                dr, dc = -1, 0
            elif dr == 1 and r == max_r:
                max_c -= 1
                dr, dc = 0, -1
            elif dr == -1 and r == min_r:
                min_c += 1
                dr, dc = 0, 1
            
            r += dr
            c += dc

        return res